src/SFML/Audio/SoundFileReaderOpus.cpp

```diff
@@ -0,0 +1,194 @@
+#include <SFML/Audio/SoundFileReaderOpus.hpp>
+#include <SFML/System/MemoryInputStream.hpp>
+#include <SFML/System/Err.hpp>
+#include <algorithm>
+#include <cctype>
+#include <cassert>
+#include <climits>
+
+namespace
+{
+    int read(void* data, unsigned char* ptr, int bytes)
+    {
+        sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+        return static_cast<int>(stream->read(ptr, bytes));
+    }
+
+    int seek(void* data, opus_int64 offset, int whence)
+    {
+        sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+        switch (whence)
+        {
+            case SEEK_SET:
+                break;
+            case SEEK_CUR:
+                offset += stream->tell();
+                break;
+            case SEEK_END:
+                offset = stream->getSize() - offset;
+        }
+
+        // Return value expected from libopusfile: 0 - Success and -1 - Failure
+        return static_cast<int>(stream->seek(offset)) >= 0 ? 0 : -1;
+    }
+
+    opus_int64 tell(void* data)
+    {
+        sf::InputStream* stream = static_cast<sf::InputStream*>(data);
+        return static_cast<opus_int64>(stream->tell());
+    }
+
+    const OpusFileCallbacks callbacks = {&read, &seek, &tell, NULL};
+}
+
+namespace sf
+{
+namespace priv
+{
+////////////////////////////////////////////////////////////
+bool SoundFileReaderOpus::check(InputStream& stream)
+{
+    int error = 0;
+    OggOpusFile* file = op_test_callbacks(&stream, &callbacks, NULL, 0, &error);
+    if (error == 0)
+    {
+        op_free(file);
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+SoundFileReaderOpus::SoundFileReaderOpus() :
+m_opus        (NULL),
+m_channelCount(0)
+{
+}
+
+
+////////////////////////////////////////////////////////////
+SoundFileReaderOpus::~SoundFileReaderOpus()
+{
+    close();
+}
+
+
+////////////////////////////////////////////////////////////
+bool SoundFileReaderOpus::open(InputStream& stream, Info& info)
+{
+    // Open the Opus stream
+    int status = 0;
+    m_opus = op_open_callbacks(&stream, &callbacks, NULL, 0, &status);
+    if (status != 0)
+    {
+        err() << "Failed to open Opus file for reading" << std::endl;
+        return false;
+    }
+
+    // Retrieve the music attributes
+    const OpusHead* opusHead = op_head(m_opus, -1);
+    info.channelCount = opusHead->channel_count;
+    info.sampleRate = opusHead->input_sample_rate;
+    info.sampleCount = static_cast<std::size_t>(op_pcm_total(m_opus, -1) * opusHead->channel_count);
+
+    // We must keep the channel count for the seek function
+    m_channelCount = info.channelCount;
+
+    return true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SoundFileReaderOpus::seek(Uint64 sampleOffset)
+{
+    assert(m_opus != NULL);
+
+    op_pcm_seek(m_opus, sampleOffset / m_channelCount);
```

Comment:
Based on the docstring for `SoundFileReader::seek`
https://github.com/SFML/SFML/blob/e31b925234676472f2705f8cfe0dc46901518929/include/SFML/Audio/SoundFileReader.hpp#L79-L91

and the docstring for [op_pcm_seek](https://opus-codec.org/docs/opusfile_api-0.7/group__stream__seeking.html#gad4f20d0601c2bab8d724f647673ae92c).

should it not also be divided by the sample rate, in this case 48000.

Reply:
I don't think so. It wouldn't be working at all if this was the case, plus Opus uses the same API as OGG (and FLAC) and there it isn't needed either, plus dividing by 48000 sound like a huge number, not surprising if this would introduce odd rounding issues.

Reply:
The function wants a number of samples, that's what we give it. Dividing by the sample rate would give a number of seconds.

Reply:
Then I don't get the message

```
/// Offsets can be calculated like this: 
/// `sampleNumber * sampleRate * channelCount` 
```

Disregarding the channelCount

sampleNumber  still has dimension [samples]
sampleRate is [samples/sec]

according SoundFileReader::seek "offset" expression I would get [samples^2/sec]. What exactly is this? I'm I making a mistake somewhere?

Reply:
This part of the documentation was added later and is obviously wrong, indeed. Feel free to remove or correct it (sampleNumber should be "timeInSeconds").